iree/turbine/kernel/wave/compile.py

```diff
@@ -0,0 +1,128 @@
+        self.subs = options.subs
+        self.executable = executable
+        self.asm = asm
+        # Load kernel from cache if available.
+        self.load_from_cache()
+
+    def load_from_cache(self):
+        cache_enabled = is_cache_enabled()
+        if cache_enabled:
+            cache_manager = get_cache_manager()
+            possible_cached_kernel = cache_manager.load_kernel(self.options.kernel_hash)
+            if possible_cached_kernel:
+                self.executable = possible_cached_kernel.vmfb
+
+    def __call__(self, *args, **kwargs):
+        return self.invoke(*args, **kwargs)
+
+    def invoke(self, *args, **kwargs):
+        """
+        Invokes the wave kernel with the given arguments.
+        Returns the assembly code of the compiled kernel.
+        """
+
+        # Partition arguments into kernel inputs and outputs.
+        kernel_inputs = []
+        kernel_outputs = []
+        for arg, b in zip(args, self.options.kernel_sig.kernel_buffer_bindings):
+            usage = b.kernel_buffer_type.usage
+            if usage == kernel_codegen.KernelBufferUsage.INPUT:
+                kernel_inputs.append(arg)
+
+            if usage == kernel_codegen.KernelBufferUsage.OUTPUT:
+                kernel_outputs.append(arg)
+
+        invoke_vmfb(self.executable, self.options, kernel_inputs, kernel_outputs)
+        return self.asm
+
+
+def wave_compile(options: WaveCompileOptions, kernel: "LaunchableWave") -> WaveKernel:
+    """
+    Compiles the wave kernel to an executable.
+    """
+
+    # Create an indexing context and populate substitutions.
+    push(IndexingContext, IndexingContext())
+    idxc = IndexingContext.current()
+    idxc.subs = options.subs
+
+    # For the wave runtime, we need the hsaco binary. So we turn on
+    # dumping of binaries and store in wave runtime directory. If we
+    # are caching, this will be moved to the appropriate directory.
+    if options.wave_runtime:
+        options.dump_binaries = str(WAVE_RUNTIME_DIR)
+
+    # Recompile kernel from scratch if not found in cache.
+    (
+        mb,
+        graph,
+        exe,
+        kernel_sig,
+        entrypoint_name,
+        options,
+    ) = kernel._trace_and_get_kernel_signature(options)
+    options.kernel_sig = kernel_sig
+
+    host_codegen.isolated_test_call(
+        mb, exe, kernel_sig, entrypoint_name, options.dynamic_symbols
+    )
+    asm = mb.module_op.get_asm()
+    if options.print_mlir:
+        print(asm)
+
+    if options.override_mlir:
+        asm = options.override_mlir
+
+    if options.compile_to_mlir:
+        return WaveKernel(options, None, asm)
+
+    compiled_wave_vmfb = compile_to_vmfb(asm, options)
+    if options.create_vmfb_file:
+        _write_file(options.create_vmfb_file, "wb", compiled_wave_vmfb)
+
+    kernel_usages = [
+        binding.kernel_buffer_type.usage
+        for binding in kernel_sig.kernel_buffer_bindings
+    ]
+
+    if is_cache_enabled():
```

Comment:
I may be missing something here, can you explain when are we "skipping" the compilation pipeline if we found it in the cache? Seems like right now it is only storing into cache

Reply:
We never skip the compilation pipeline. Now that compilation and invocations are separated, we compile only once and invoke multiple times. Does that make sense?

Reply:
1. is resolved with LRU caching at sglang level
2. will be resolved by moving `load_from_cache` into start of wave_compile